readme.md
# nbt2json

A command line utility and Go module that reads NBT data and converts it to JSON or YAML for editing and then back to NBT.

## Features

- nbt2json executable will auto-detect and decompress gzipped files
- nbt2json executable has option to gzip output
- Can read and write both Minecraft Bedrock Edition and Java Edition NBT data
    - Does **not** auto-detect which
    - nbt2json executable defaults to Bedrock Edition / little endian
- Can import to other Go projects
- Can use either JSON or YAML
- Can include comment in JSON/YAML output (which is ignored when converting back to NBT)

## Help screen

By defualt, the nbt2json executable waits for input from stdin, so you need to `nbt2json -h` to see the help screen.

```
NAME:
   NBT to JSON - Converts NBT-encoded data to JSON | https://github.com/midnightfreddie/nbt2json

USAGE:
   nbt2json.exe [global options] command [command options] [arguments...]

VERSION:
   0.4.0

AUTHOR:
   Jim Nelson <jim@jimnelson.us>

COMMANDS:
   help, h  Shows a list of commands or help for one command

GLOBAL OPTIONS:
   --reverse, -r                  Convert JSON to NBT instead (default: false)
   --gzip, -z                     Compress output with gzip (default: false)
   --comment COMMENT, -c COMMENT  Add COMMENT to json or yaml output, use quotes if contains white space
   --big-endian, --java, -b       Use for Minecraft Java Edition (like most other NBT tools) (default: false)
   --in FILE, -i FILE             Input FILE path (default: "-")
   --out FILE, -o FILE            Output FILE path (default: "-")
   --yaml, --yml, -y              Use YAML instead of JSON (default: false)
   --long-as-string, -l           If set, nbt long values will be a string instead of uint32 pair (default: false)
   --skip NUM                     Skip NUM bytes of NBT input. For Bedrock's level.dat, use --skip 8 to bypass header (default: 0)
   --help, -h                     show help (default: false)
   --version, -v                  print the version (default: false)

COPYRIGHT:
   (c) 2018, 2019, 2020 Jim Nelson
```

## JSON format notes

The JSON document should be an object with an `"nbt"` field which is an array.
Objects in the array represent one tag and will typically have tagType, name,
and value fields. A typical Minecraft NBT will have one compound tag with a
number of other tags inside that one. This converter needs at least one tag.

Many JSON libraries cannot properly handle a 64-bit integer, so nbt2json handles
the long tag in one of two special ways for portability and compatibility.

By default it presents valueLeast and valueMost unsigned 32-bit integers.
valueMost is the most significant 32 bits, so it should be rolled bitwise to the
left 32 bits in a 64-bit space and added (or bitwise 'or'ed) to valueLeast to
reassemble the 64-bit long singed integer. Here is an example:

```json
{
  "nbt": [
    {
      "tagType": 4,
      "name": "LongAsUint32PairExample",
      "value": {
        "valueLeast": 4294967295,
        "valueMost": 2147483647
      }
    }
  ]
}
```

Alternately it can turn the 64-bit integer into a string. This would be easier
to edit by hand in a text editor. Example:

```json
{
  "nbt": [
    {
      "tagType": 4,
      "name": "LongAsStringExample",
      "value": "9223372036854775807"
    }
  ]
}
```

## Dev notes

- Client Go code needs to `import "github.com/midnightfreddie/nbt2json"`
- Defaults to little-endian encoding for Bedrock Edition. Call `nbt2json.UseJavaEncoding()` and `nbt2json.UseBedrockEncoding()` to change encoding mode for as long as the module is open.
- The functions use byte arrays where you might expect strings. Convert as such: `var myString = someByteArray[:]` or `var myByteArray = []byte(someStringValue)`
- All errors should bubble up through the error part of the result and should describe where the problem was
- The Json2Nbt function uses an `interface{}` and encodes based on the tagType fields. I had originally hoped to Marshal and Unmarshal to and from JSON and NBT, but my goal was to export to JSON, edit and then reencode. This way the struct doesn't have to match the data schema.
- My main motivation for this project is to convert to/from JSON and use any JSON editor to modify Minecraft PE data with [McpeTool](https://github.com/midnightfreddie/McpeTool), and to keep the read/write primitives in Go code while letting a client browser manage any validation to avoid having to re-release the read/write tools every time Minecraft changes formats.

### Exported Go functions

- **Nbt2Yaml** converts uncompressed NBT byte array to YAML byte array

		func Nbt2Yaml(b []byte, comment string) ([]byte, error)

- **Nbt2Json** converts uncompressed NBT byte array to JSON byte array

		func Nbt2Json(b []byte, comment string) ([]byte, error)

- **Yaml2Nbt** converts JSON byte array to uncompressed NBT byte array (Hint: You can just use this for both JSON *and* YAML if you like since JSON is a valid subeset of YAML)

		func Yaml2Nbt(b []byte) ([]byte, error)

- **Json2Nbt** converts JSON byte array to uncompressed NBT byte array

		func Json2Nbt(b []byte) ([]byte, error)

- **UseJavaEndoding** sets any nbt encoding/decoding to big-endian to match Minecraft Java Edition

        func UseJavaEncoding()

- **UseBedrockEncoding** sets nbt encoding/decoding to little-endian to match Minecraft Bedrock Edition (default)

        func UseBedrockEncoding()

- **UseLongAsString** sets json output for nbt long tags to numbers in strings

        func UseLongAsString()

- **UseLongAsUint32Pair** sets json output for nbt long tags to a uint32 pair (default, needed for json portability)

        func UseLongAsUint32Pair()

Other exports of possible interest are in common.go.


changelog.md
## v0.4.0

Breaking changes!

For everyone:

- NBT Long values are now stored as valueLeast & valueMost **32-bit unsigned
integer pairs** in JSON. This is to prevent conversion issues across various
languages' JSON libaries, but it will make a little more work if you need to
modify these values.
  - Example JSON for NBT long

        {
          "nbt": [
            {
              "tagType": 4,
              "name": "LongAsUint32Pair",
              "value": {
                "valueLeast": 4294967295,
                "valueMost": 2147483647
              }
            }
          ]
        }

  - Example JSON for NBT long as string

        {
          "nbt": [
            {
              "tagType": 4,
              "name": "LongAsString",
              "value": "9223372036854775807"
            }
          ]
        }

- Optionally you can store long values as strings in JSON which may be more
convenient depending on the use case
- Converting from JSON will accept either strings or the valueLeast/valueMost
pair automatically

For utility executable users:

- Some parameters have been removed or renamed
  - Since Bedrock is default, all the little-endian parameters are gone
  - `--json2nbt`, `--big`, and `--pc` were removed, but their aliases remain
  - Command line library has been updated to the latest version
  - Added `--long-as-string` and short alias `-l` to cause NBT long values
  (64-bit integers) to be numbers-in-strings in the JSON output instead of
  valueLeast & valueMost numbers

For devs:

- byteOrder arguments are gone from all functions
- Use `UseJavaEncoding()` if you need big-endian / Java Edition encoding.
`UseBedrockEncoding()` is set by default, but you can call it to switch back if
you switched to Java previously.
- Use `UseLongAsString()` if you want NBT long int64's to be in the JSON as
strings. `UseLongAsUint32Pair()` is set by default, but you can call it to
switch back if you set the string method previously.
- Go tests are more thorough

## v0.3.4

This version has no data differences from v.0.3.3. It just has improved error
messaging and Go tests. I wanted to make this release mostly for those who would
use it as a module so they have the new improvements before I make some breaking
changes for the next version.

- Added Go tests in nbt2json_test.go
- Corrected mislabeled error messages
- Added errors for missing "nbt" in json and out-of-range numbers
- Added input value to error message output
- Refactored module code to include commmon.go for possibly interesting exports
- Executable now pulls version and url from the module values in common.go

## v0.3.3

- Merged VADemon's fix for int64s
- Int64s (NBT "long"s) will now export and import properly; they were previously getting messed up which impacted UUIDs as well
- However, other languages' JSON libraries may not handle int64s in JSON; this will be addressed in future release

## v0.3.2

- Fixed emtpy array showing as `null` in json instead of `[]` for compound tag 10
- Moved nbt2json/ main executable to cmd/nbt2json/
- Added `--gzip` / `-z` option to executable to compress output

There does not seem to be a missing data type, but users are reporting UUIDs
are not preserving due to float interpretation of json numbers.

## v0.3.1

- Added go.mod and go.sum to help manage dependency versions

There is ~~at least one new NBT data type and~~ one bug since the last update,
but these are not yet addressed. I am just adding go.mod to prevent possible
dependenccy version issues. I hope to address the issues soon.

nbt2json.go
package nbt2json

import (
	"bytes"
	"encoding/binary"
	"encoding/json"
	"fmt"
	"math"
	"time"

	"github.com/ghodss/yaml"
)

// NbtJson is the top-level JSON document; it is exported for reflect, and client code shouldn't use it
type NbtJson struct {
	Name           string             `json:"name"`
	Version        string             `json:"version"`
	Nbt2JsonUrl    string             `json:"nbt2JsonUrl"`
	ConversionTime string             `json:"conversionTime,omitempty"`
	Comment        string             `json:"comment,omitempty"`
	Nbt            []*json.RawMessage `json:"nbt"`
}

// NbtTag represents one NBT tag for each struct; it is exported for reflect, and client code shouldn't use it
type NbtTag struct {
	TagType byte        `json:"tagType"`
	Name    string      `json:"name"`
	Value   interface{} `json:"value,omitempty"`
}

// NbtTagList represents an NBT tag list; it is exported for reflect, and client code shouldn't use it
type NbtTagList struct {
	TagListType byte          `json:"tagListType"`
	List        []interface{} `json:"list"`
}

// NbtLong stores a 64-bit int into two 32-bit values for json portability. ValueMost are the high 32 bits and ValueLeast are the low 32 bits.
//   using uint32s to avoid Go trying to outsmart us on "negative" int32s
type NbtLong struct {
	ValueLeast uint32 `json:"valueLeast"`
	ValueMost  uint32 `json:"valueMost"`
}

// Turns an int64 (nbt long) into a valueLeast/valueMost json pair
func longToIntPair(i int64) NbtLong {
	var nbtLong NbtLong
	nbtLong.ValueLeast = uint32(i & 0xffffffff)
	nbtLong.ValueMost = uint32(i >> 32)
	return nbtLong
}

func intPairToLong(nbtLong NbtLong) int64 {
	var i int64
	i = int64(nbtLong.ValueLeast) | (int64(nbtLong.ValueMost) << 32)
	return i
}

// Nbt2Yaml converts uncompressed NBT byte array to YAML byte array
func Nbt2Yaml(b []byte, comment string) ([]byte, error) {
	jsonOut, err := Nbt2Json(b, comment)
	if err != nil {
		return nil, err
	}
	yamlOut, err := yaml.JSONToYAML(jsonOut)
	if err != nil {
		return yamlOut, NbtParseError{"Error converting JSON to YAML. Oops. JSON conversion succeeded, so please report this error and use JSON instead.", err}
	}
	return yamlOut, nil
}

// Nbt2Json converts uncompressed NBT byte array to JSON byte array
func Nbt2Json(b []byte, comment string) ([]byte, error) {
	var nbtJson NbtJson
	nbtJson.Name = Name
	nbtJson.Version = Version
	nbtJson.Nbt2JsonUrl = Nbt2JsonUrl
	nbtJson.ConversionTime = time.Now().Format(time.RFC3339)
	nbtJson.Comment = comment
	buf := bytes.NewReader(b)
	// var nbtJson.nbt []*json.RawMessage
	for buf.Len() > 0 {
		element, err := getTag(buf)
		if err != nil {
			return nil, err
		}
		myTemp := json.RawMessage(element)
		nbtJson.Nbt = append(nbtJson.Nbt, &myTemp)
	}
	jsonOut, err := json.MarshalIndent(nbtJson, "", "  ")
	if err != nil {
		return nil, err
	}
	return jsonOut, nil
}

// getTag broken out form Nbt2Json to allow recursion with reader but public input is []byte
func getTag(r *bytes.Reader) ([]byte, error) {
	var data NbtTag
	err := binary.Read(r, byteOrder, &data.TagType)
	if err != nil {
		return nil, NbtParseError{"Reading TagType", err}
	}
	// do not try to fetch name for TagType 0 which is compound end tag
	if data.TagType != 0 {
		var err error
		var nameLen int16
		err = binary.Read(r, byteOrder, &nameLen)
		if err != nil {
			return nil, NbtParseError{"Reading Name length", err}
		}
		name := make([]byte, nameLen)
		err = binary.Read(r, byteOrder, &name)
		if err != nil {
			return nil, NbtParseError{fmt.Sprintf("Reading Name - is UseJavaEncoding or UseBedrockEncoding set correctly? Name length decoded is %d", nameLen), err}
		}
		data.Name = string(name[:])
	}
	data.Value, err = getPayload(r, data.TagType)
	if err != nil {
		return nil, err
	}
	outJson, err := json.MarshalIndent(data, "", "  ")
	return outJson, err
}

// Gets the tag payload. Had to break this out from the main function to allow tag list recursion
func getPayload(r *bytes.Reader, tagType byte) (interface{}, error) {
	var output interface{}
	var err error
	switch tagType {
	case 0:
		// end tag for compound; do nothing further
	case 1:
		var i int8
		err = binary.Read(r, byteOrder, &i)
		if err != nil {
			return nil, NbtParseError{"Reading int8", err}
		}
		output = i
	case 2:
		var i int16
		err = binary.Read(r, byteOrder, &i)
		if err != nil {
			return nil, NbtParseError{"Reading int16", err}
		}
		output = i
	case 3:
		var i int32
		err = binary.Read(r, byteOrder, &i)
		if err != nil {
			return nil, NbtParseError{"Reading int32", err}
		}
		output = i
	case 4:
		var i int64
		err = binary.Read(r, byteOrder, &i)
		if err != nil {
			return nil, NbtParseError{"Reading int64", err}
		}
		if longAsString {
			output = fmt.Sprintf("%d", i)
		} else {
			output = longToIntPair(i)
		}
	case 5:
		var f float32
		err = binary.Read(r, byteOrder, &f)
		if err != nil {
			return nil, NbtParseError{"Reading float32", err}
		}
		output = f
	case 6:
		var f float64
		err = binary.Read(r, byteOrder, &f)
		if err != nil {
			return nil, NbtParseError{"Reading float64", err}
		}
		if math.IsNaN(f) {
			output = "NaN"
		} else {
			output = f
		}
	case 7:
		var byteArray []int8
		var oneByte int8
		var numRecords int32
		err := binary.Read(r, byteOrder, &numRecords)
		if err != nil {
			return nil, NbtParseError{"Reading byte array tag length", err}
		}
		for i := int32(1); i <= numRecords; i++ {
			err = binary.Read(r, byteOrder, &oneByte)
			if err != nil {
				return nil, NbtParseError{"Reading byte in byte array tag", err}
			}
			byteArray = append(byteArray, oneByte)
		}
		output = byteArray
	case 8:
		var strLen int16
		err := binary.Read(r, byteOrder, &strLen)
		if err != nil {
			return nil, NbtParseError{"Reading string tag length", err}
		}
		utf8String := make([]byte, strLen)
		err = binary.Read(r, byteOrder, &utf8String)
		if err != nil {
			return nil, NbtParseError{"Reading string tag data", err}
		}
		output = string(utf8String[:])
	case 9:
		var tagList NbtTagList
		err = binary.Read(r, byteOrder, &tagList.TagListType)
		if err != nil {
			return nil, NbtParseError{"Reading TagType", err}
		}
		var numRecords int32
		err := binary.Read(r, byteOrder, &numRecords)
		if err != nil {
			return nil, NbtParseError{"Reading list tag length", err}
		}
		for i := int32(1); i <= numRecords; i++ {
			payload, err := getPayload(r, tagList.TagListType)
			if err != nil {
				return nil, NbtParseError{"Reading list tag item", err}
			}
			tagList.List = append(tagList.List, payload)
		}
		output = tagList
	case 10:
		var compound []json.RawMessage
		var tagType byte
		for err = binary.Read(r, byteOrder, &tagType); tagType != 0; err = binary.Read(r, byteOrder, &tagType) {
			if err != nil {
				return nil, NbtParseError{"compound: reading next tag type", err}
			}
			_, err = r.Seek(-1, 1)
			if err != nil {
				return nil, NbtParseError{"seeking back one", err}
			}
			tag, err := getTag(r)
			if err != nil {
				return nil, NbtParseError{"compound: reading a child tag", err}
			}
			compound = append(compound, json.RawMessage(string(tag)))
		}
		if compound == nil {
			// Explicitly give empty array else value will be null instead of []
			output = []int{}
		} else {
			output = compound
		}
	case 11:
		var intArray []int32
		var numRecords, oneInt int32
		err := binary.Read(r, byteOrder, &numRecords)
		if err != nil {
			return nil, NbtParseError{"Reading int array tag length", err}
		}
		for i := int32(1); i <= numRecords; i++ {
			err := binary.Read(r, byteOrder, &oneInt)
			if err != nil {
				return nil, NbtParseError{"Reading int in int array tag", err}
			}
			intArray = append(intArray, oneInt)
		}
		output = intArray
	case 12:
		var longArray []NbtLong
		var longStringArray []string
		var numRecords, oneInt int64
		err := binary.Read(r, byteOrder, &numRecords)
		if err != nil {
			return nil, NbtParseError{"Reading long array tag length", err}
		}
		for i := int64(1); i <= numRecords; i++ {
			err := binary.Read(r, byteOrder, &oneInt)
			if err != nil {
				return nil, NbtParseError{"Reading long in long array tag", err}
			}
			longArray = append(longArray, longToIntPair(i))
			longStringArray = append(longStringArray, fmt.Sprintf("%d", i))
		}
		if longAsString {
			output = longStringArray
		} else {
			output = longArray
		}

	default:
		return nil, NbtParseError{fmt.Sprintf("TagType %d not recognized", tagType), nil}
	}
	return output, nil
}
